### PR TITLE
Add audit tool to extract inconsistencies between users and buckets

### PR DESCRIPTION
**Pull request by shino**
There is possibility of inconsistencies between the user bucket `moss.users`
and the buckets bucket `moss.buckets` because no transaction is (can be) used.

This PR adds tool for detecting inconsistencies between them.
Three kinds of inconsistencies:
1. Bucket object in buckets bucket is not tracked by the user
   in its value. (`{not_tracked, {User, Bucket}}`)
2. A certain user thinks it owns a bucket but no objects in buckets bucket
   exists. (`{no_bucket_object, {User, Bucket}}`)
3. A certain user thinks it owns a bucket but the bucket is actually
   owned by different user. (`{different_user, {User, Bucket, DifferentUser}}`)

---

Sample execution by console command and log output in it:

```
% dev/dev1/bin/riak-cs-admin audit-bucket-ownership
[{no_bucket_object,{<<"T_RNSHIN22TIJ1MKF8BP">>,<<"corrupt">>}},
 {different_user,{<<"N4ADKYOVGZLKL5RVD0MA">>,<<"bob2">>,
                  <<"T_RNSHIN22TIJ1MKF8BP">>}},
 {not_tracked,{<<"T_RNSHIN22TIJ1MKF8BP">>,<<"missing-in-list">>}},
 {not_tracked,{<<"T_RNSHIN22TIJ1MKF8BP">>,<<"bob2">>}},
 {not_tracked,{<<"N4ADKYOVGZLKL5RVD0MA">>,<<"missing-in-bobs-list">>}}]
```

```
16:54:20.475 [info] Bucket is not tracked by user: {User, Bucket} = {N4ADKYOVGZLKL5RVD0MA, missing-in-bobs-list}
16:54:20.475 [info] Bucket is not tracked by user: {User, Bucket} = {T_RNSHIN22TIJ1MKF8BP, bob2}
16:54:20.475 [info] Bucket is not tracked by user: {User, Bucket} = {T_RNSHIN22TIJ1MKF8BP, missing-in-list}
16:54:20.475 [info] Bucket is owned by different user: {User, Bucket, DifferentUser} = {N4ADKYOVGZLKL5RVD0MA, bob2, T_RNSHIN22TIJ1MKF8BP}
16:54:20.475 [info] Bucket does not exist: {User, Bucket} = {T_RNSHIN22TIJ1MKF8BP, corrupt}
```

---

This PR is in the state of request-for-comment.  Any comments on categorization,
log format, command line output, implementation, etc is appreciated.

### CODE REVIEW

**Comment by kuenishi** [comment]
Note: this work is for #1161 (RCS-211) .

File: src/riak_cs_console.erl [hunk 1]
```diff
@@ -62,6 +63,86 @@ cluster_info([OutFile]) ->
...
+%% @doc Audit bucket ownership by comparing information between
+%%      users bucket and buckets bucket.
+audit_bucket_ownership(_Args) when is_list(_Args) ->
+    audit_bucket_ownership().
+
+audit_bucket_ownership() ->
+    {ok, RcPid} = riak_cs_riak_client:start_link([]),
+    try
+        Inconsistencies = audit_bucket_ownership0(RcPid),
+        io:format("~p", [Inconsistencies])
```
Comment (kuenishi) [inline src/riak_cs_console.erl]: `~n` would make this look better. Also, a message like "No inconsistency found." will be more informative when `Inconsistencies` is just `[]`.

**Comment by kuenishi** [comment]
It worked;

``` sh
$ curl -XPUT  'http://localhost:8098/buckets/moss.buckets/keys/test' -H "X-Riak-Vclock: a85hYGBgzGDKBVI8ypz/fp7QY+KCCCUy5rEyKCjLnOfLAgA=" -d "sub-key"
$ rel/riak-cs/bin/riak-cs-admin audit-bucket-ownership
[{different_user,{<<"admin-key">>,<<"test">>,<<"sub-key">>}},
 {not_tracked,{<<"sub-key">>,<<"test">>}}]
```

We would follow this with some pretty printing or documentation.

File: src/riak_cs_console.erl [hunk 2]
```diff
@@ -62,6 +63,86 @@ cluster_info([OutFile]) ->
...
+audit_bucket_ownership0(RcPid) ->
+    FromUsers = ownership_from_users(RcPid),
+    {FromBuckets, OwnedBy} = ownership_from_buckets(RcPid),
+    lager:debug("FromUsers: ~p~n", [lists:usort(gb_sets:to_list(FromUsers))]),
+    lager:debug("FromBuckets: ~p~n", [lists:usort(gb_sets:to_list(FromBuckets))]),
+    lager:debug("OwnedBy: ~p~n", [lists:usort(gb_trees:to_list(OwnedBy))]),
+    Inconsistencies0 =
+        gb_sets:fold(
+          fun ({U, B}, Acc) ->
+                  lager:info(
```
Comment (kuenishi) [inline src/riak_cs_console.erl:93]: It would also be nice to output console.

File: src/riak_cs_console.erl [hunk 3]
```diff
@@ -62,6 +63,86 @@ cluster_info([OutFile]) ->
...
+                    "Bucket is not tracked by user: {User, Bucket} = {~s, ~s}",
+                    [U, B]),
+                  [{not_tracked, {U, B}} | Acc]
+          end, [], gb_sets:subtract(FromBuckets, FromUsers)),
+    gb_sets:fold(
+      fun({U,B}, Acc) ->
+              case gb_trees:lookup(B, OwnedBy) of
+                  none ->
+                      lager:info(
```
Comment (kuenishi) [inline src/riak_cs_console.erl:102]: Ditto.

File: src/riak_cs_console.erl [hunk 4]
```diff
@@ -62,6 +63,86 @@ cluster_info([OutFile]) ->
...
+                        "Bucket does not exist: {User, Bucket} = {~s, ~s}", [U, B]),
+                      [{no_bucket_object, {U, B}} | Acc];
+                  {value, DifferentUser} ->
+                      lager:info(
```
Comment (kuenishi) [inline src/riak_cs_console.erl:106]: Ditto.

File: src/riak_cs_console.erl (hunk 2 above)
Comment (shino) [inline src/riak_cs_console.erl:93]: By discussion in chat, output to console may be very noisy when large number of inconsistencies.

**Comment by shino** [comment]
Thanks for review :smile: Updated.

**Comment by shino** [comment]
@borshop merge

**Comment by shino** [comment]
For release note: please look at the description at the top.

**Comment by shino** [comment]
micro benchmark with 100,000 users and 27,000 buckets and small inconsistencies:

```
% time ./dev/dev1/erts-5.10.3/bin/nodetool -name 'rcs-dev1@127.0.0.1' -setcookie riak rpc_infinity riak_cs_console audit_bucket_ownership
[{no_bucket_object,{<<"T_RNSHIN22TIJ1MKF8BP">>,<<"corrupt">>}},
 {different_user,{<<"N4ADKYOVGZLKL5RVD0MA">>,<<"bob2">>,
                  <<"T_RNSHIN22TIJ1MKF8BP">>}},
 {not_tracked,{<<"T_RNSHIN22TIJ1MKF8BP">>,<<"missing-in-list">>}},
 {not_tracked,{<<"T_RNSHIN22TIJ1MKF8BP">>,<<"bob2">>}},
 {not_tracked,{<<"N4ADKYOVGZLKL5RVD0MA">>,<<"missing-in-bobs-list">>}}].
./dev/dev1/erts-5.10.3/bin/nodetool -name 'rcs-dev1@127.0.0.1' -setcookie ria  0.28s user 0.11s system 0% cpu 2:18.44 total
```

**Comment by shino** [comment]
In local environment (single node cluster with ring size 8, not good for benchmark), it took a few seconds for list keys, it seems that getting every object in serialized way is current bottleneck. But, one more point, Riak CS node consumes CPU ~ 300%. Feels strange. Further investigation / optimization may follow in another ticket / issue.

**Pull request merged**